File: lab5 heap/D.py
class MaxHeap:
    def __init__(self):
        self.a = []
        self.pos = int()

    def parent(self, i):
        return (i - 1) // 2

    def left(self, i):
        return i * 2 + 1

    def right(self, i):
        return i * 2 + 2

    def getMax(self):
        return self.a[0]

    def heapify(self, i):
        if self.left(i) > len(self.a) - 1:
            self.pos = i
            return
        j = self.left(i)
        if self.right(i) < len(self.a) and self.a[self.left(i)] < self.a[self.right(i)]:
            j = self.right(i)
        if self.a[i] < self.a[j]:
            self.a[i], self.a[j] = self.a[j], self.a[i]
            self.pos = j
            self.heapify(j)

    def insert(self, value):
        self.a.append(value)
        i = len(self.a) - 1
        while i > 0 and self.a[self.parent(i) < self.a[i]]:
            self.a[self.parent(i)], self.a[i] = self.a[i], self.a[self.parent(i)]
            i = self.parent(i)
        return i

    def extractMax(self):
        root = self.getMax()
        self.a[0], self.a[-1] = self.a[-1], self.a[0]
        self.a.pop(-1)
        self.pos = 0
        self.heapify(0)
        return root


heap = MaxHeap()
n, m = [int(i) for i in input().split()]
for i in range(m):
    l = [int(i) for i in input().split()]

    if l[0] == 1:
        if len(heap.a) == 0:
            print(-1)
        elif len(heap.a) == 1:
            print(0, heap.extractMax())
        else:
            res = heap.extractMax()
            print(heap.pos + 1, res)
    elif l[0] == 2:
        if len(heap.a) >= n:
            print(-1)
        else:
            print(heap.insert(l[1]) + 1)

    for i in range(len(heap.a)):
        print(heap.a[i], end=' ')

File: lab5 heap/D.cpp
#include <iostream>
 
using namespace std;

int heap[500000];

int heapify(int i, int sz) {
    int j = heap[i];
    int ind = i;
    if(i * 2 <= sz && heap[2 * i] > j) {
        j = heap[2 * i];
        ind = i * 2;
    }

    if(i * 2 + 1 <= sz && heap[2 * i + 1] > j) {
        j = heap[2 * i + 1];
        ind = i * 2 + 1;
    }
 
    if(ind == i)
        return i;
    swap(heap[ind], heap[i]);
    return heapify(ind, sz);
}

int SiftUp(int i) {
    while(i > 1 && heap[i] > heap[i / 2]) {
        swap(heap[i], heap[i / 2]);
        i = i / 2;
    }
    return i;
}

void insert(int x, int sz) {
    sz++;
    heap[sz] = x;
    cout << SiftUp(sz) << endl;
}

void extractMax(int sz) {
    int root = heap[1];
    swap(heap[1], heap[sz]);
    sz--;
    cout << heapify(1, sz) << ' ' << root << endl;
} 

int main () {
    int n, m;
    cin >> n >> m;
    int sz = 0;
    for (int i = 0; i < m; i++){
        int x; cin >> x; 
        if(x == 1){
            if(sz == 1){
                cout << 0 << ' ' << heap[sz] << endl;
                sz--;
            }
            else if(sz == 0) {
                cout << -1 << endl;
            }
            else {
                extractMax(sz);
                sz--;
            }
        }
        else {
            int k; cin >> k;
            if(sz >= n)
                cout << -1 << endl;
            else {
                insert(k, sz);
                sz++;
            }
        }
    }
    for(int i = 1; i <= sz; i++)
        cout << heap[i] << ' ';
    
    return 0;
}